woocommerce/woocommerce-ios
# [Bundle products in order form] Show existing bundled order item configuration and enforce quantity rules

Part of #10428 
⚠️ Please review https://github.com/woocommerce/woocommerce-ios/pull/11004 first


## Why


When a bundle order item already includes any configured bundled items, we want to pre-populate the existing configuration on the Configure screen. In addition, we want to enforce the min and optional max quantity rules in the UI so that the quantity is always valid when submitting the changes remotely (passing any invalid configuration results in silent failure without any bundled order items). Some minor polishes on the Configure screen are also included in this PR to avoid too many dependent PRs.

## How

For bundle order item's configured bundled items, they also have their corresponding order items in the order with the bundle product order item as their parent. We can filter the order items to find these child items, then filter by the product ID for the bundle item configuration view model.

For enforcing the quantity rules, `minimumQuantity: Decimal` and `maximumQuantity: Decimal?` were added to the `ProductRowViewModel` initializer with default values to maintain pre-existing use cases. I was hoping the product quantity stepper could use its own view model, but didn't want to make many more changes than necessary at this point.

Configure screen polishes:
- In `ConfigurableBundleItemView`, updated the toggle to checkbox following the Android implementation pe5pgL-3Vt-p2
- In `ProductRow`, two text labels that contain secondary information are only shown when the content is non-empty

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items.

- Go to the Orders tab
- Tap on an existing editable order
- Tap `Edit`
- Tap `Add Products` and select a bundle product --> the bundle product should be added to the order
- Tap on the expand  🔽 CTA on the bundle product order item
- Tap `Configure` --> after a bit (products being loaded remotely), the configurable bundle items should be shown
- Try incrementing/decrementing the quantity of a bundled item --> the allowed quantity range should match the bundle item settings in core (max quantity is optional)
- For the bundle item that is variable, tap `Select variation` and select a variation that doesn't have any "Any" values
- Tap `Done` --> the Configure modal should be dismissed, and the order's line items should be updated with the newly configured bundled items
- Tap `Edit` to the order form again
- Tap on the expand  🔽 CTA on the bundle product order item
- Tap `Configure` --> after a bit, the default configuration should match the existing bundled order items

## Screenshots


<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/7f3ee04b-b861-4bb3-9629-affbe60991c8" width="300" />


---
